code/game/objects/structures/coatrack.dm

```diff
@@ -0,0 +1,124 @@
+/obj/structure/coatrack
+	name = "coat rack"
+	desc = "Rack that holds coats, or hats, if you're so inclined."
+	icon = 'icons/obj/coatrack.dmi'
+	icon_state = "coatrack"
+	var/obj/item/clothing/coat
+	var/obj/item/clothing/head/hat
+	var/list/custom_sprites = list(/obj/item/clothing/head/beret/security, /obj/item/clothing/accessory/poncho/tajarancloak) // Custom manual sprite override.
+
+/obj/structure/coatrack/attack_hand(mob/user as mob)
+	if(!ishuman(user))
+		return
+	if(user.incapacitated())
+		return
+	if(!user.can_use_hand())
+		return
+	if(coat)
+		user.visible_message("[user] takes [coat] off \the [src].", SPAN_NOTICE("You take [coat] off the \the [src]."))
+		user.put_in_hands(coat)
+		coat = null
+		update_icon()
+	else if(hat)
+		user.visible_message("[user] takes [hat] off \the [src].", SPAN_NOTICE("You take [hat] off the \the [src]."))
+		user.put_in_hands(hat)
+		hat = null
+		update_icon()
+
+/obj/structure/coatrack/attackby(obj/item/W as obj, mob/user as mob)
```

Comment:
You should use a use_check / use_check_and_message proc here possibly with optional flags.